DataStructures/Graphs/A_Star.java

```diff
@@ -0,0 +1,162 @@
+            this.addEdge(new Edge(18, 8, 92));  this.addEdge(new Edge(8, 11, 87));
+
+        /*
+        .x. node
+        (y) cost
+        - or | or / bidirectional connection
+
+                              ( 98)- .7. -(86)- .4.
+                                |
+                        ( 85)- .17. -(142)- .18. -(92)- .8. -(87)- .11.
+                          |
+                         . 1. -------------------- (160)
+                          |  \                       |
+                        (211) \                     .6.
+                          |    \                     |
+                         . 5.  (101)-.13. -(138)   (115)
+                          |           |     |     /
+                        ( 99)       ( 97)   |    /
+                          |           |     |   /
+            .12. -(151)- .15. -(80)- .14.   |  /
+             |            |           |     | /
+           ( 71)        (140)       (146)- .2. -(120)
+             |            |                       |
+            .19. -( 75)- . 0.        .10. -(75)- .3.
+                          |            |
+                        (118)        ( 70)
+                          |            |
+                         .16. -(111)- .9.
+         */
+        }
+    }
+
+    private static class Edge {
+        private int from;
+        private int to;
+        private int weight;
+
+        public Edge(int from, int to, int weight) {
+            this.from = from;
+            this.to = to;
+            this.weight = weight;
+        }
+
+        public int getFrom() { return from; }
+
+        public int getTo() { return to; }
+
+        public int getWeight() { return weight; }
+    }
+
+    //class to iterate during the algorithm execution, and also used to return the solution.
+    private static class PathAndDistance {
+        private int distance; //distance advanced so far.
+        private ArrayList<Integer> path; //list of visited nodes in this path.
+        private int estimated; //heuristic value associated to the last node od the path (current node).
+
+        public PathAndDistance(int distance, ArrayList<Integer> path, int estimated) {
+            this.distance = distance;
+            this.path = path;
+            this.estimated = estimated;
+        }
+
+        public int getDistance() { return distance; }
+
+        public ArrayList<Integer> getPath() { return path; }
+
+        public int getEstimated() { return estimated; }
+
+        private void printSolution () {
+            if (this.path != null)
+                System.out.println("Optimal path: " + this.path.toString() +
+                        ", distance: " + this.distance);
+            else
+                System.out.println("There is no path available to connect the points");
+        }
+    }
+
+    public static void main(String[] args) {
+        //heuristic function optimistic values
+        int[] heuristic = {366, 0, 160, 242, 161, 178, 77, 151, 226,
```

Comment:
What does this array represent?

Reply:
The estimated values calculated to reach point x from point y. This algorithm needs a collection of values like this to make decisions. It's up to the user the way in which those values are calculated.